solutions/day-01/exercise3/level3/level3.js

```diff
@@ -0,0 +1,195 @@
+    isLoggedIn: true,
+  },
+  {
+    _id: "eefamr",
+    username: "Martha",
+    email: "martha@martha.com",
+    password: "123222",
+    createdAt: "08/01/2020 9:50 AM",
+    isLoggedIn: false,
+  },
+  {
+    _id: "ghderc",
+    username: "Thomas",
+    email: "thomas@thomas.com",
+    password: "123333",
+    createdAt: "08/01/2020 10:00 AM",
+    isLoggedIn: false,
+  },
+];
+
+const products = [
+  {
+    _id: "eedfcf",
+    name: "mobile phone",
+    description: "Huawei Honor",
+    price: 200,
+    ratings: [
+      { userId: "fg12cy", rate: 5 },
+      { userId: "zwf8md", rate: 4.5 },
+    ],
+    likes: [],
+  },
+  {
+    _id: "aegfal",
+    name: "Laptop",
+    description: "MacPro: System Darwin",
+    price: 2500,
+    ratings: [],
+    likes: ["fg12cy"],
+  },
+  {
+    _id: "hedfcg",
+    name: "TV",
+    description: "Smart TV:Procaster",
+    price: 400,
+    ratings: [{ userId: "fg12cy", rate: 5 }],
+    likes: ["fg12cy"],
+  },
+];
+
+// 2. Imagine you are getting the above users collection from a MongoDB database.
+
+// 2.1. Create a function called signUp which allows user to add to the collection.
+// If user exists, inform the user that he has already an account.
+const signUp = (newUser) => {
+  const doesExist = users.some(
+    (user) => user.email.toLowerCase() === newUser.email.toLowerCase()
+  );
+  if (doesExist) {
+    console.log("You have already an account.");
+  } else {
+    users.push(newUser);
+  }
+};
+
+signUp({
+  _id: "ghderc",
+  username: "Thomas",
+  email: "thomas@thomas.com",
+  password: "123333",
+  createdAt: "08/01/2020 10:00 AM",
+  isLoggedIn: false,
+});
+
+// 2.2 Create a function called signIn which allows user to sign in to the application
+const signIn = (user) => {
+  const isValid = users.some(
+    (storedUser) =>
+      storedUser.email.toLowerCase() === user.email.toLowerCase() &&
+      storedUser.password === user.password
+  );
+  if (isValid) {
+    console.log("Sign-in successful.");
+  } else {
+    console.log("Invalid email or password!");
+  }
+};
+
+signIn({
+  email: "thomas@thomas.com",
+  password: "123333",
+});
+
+// 3. The products array has three elements and each of them has six properties.
+
+// 3.1 Create a function called rateProduct which rates the product
+const rateProduct = (productId, userId, rating) => {
+  const product = products.find((product) => product._id === productId);
+  if (product) {
+    product.ratings.push({ userId, rate: rating });
+  }
+};
+
+rateProduct("eedfcf", "xnxx", 5);
+
+// 3.2 Create a function called averageRating which calculate the average rating of a product
+const averageRating = (productId) => {
+  const product = products.find((product) => product._id === productId);
+  if (product) {
+    const ratings = product.ratings.map((rating) => rating.rate);
+    const total = ratings.reduce((total, rating) => total + rating, 0);
+    return total / ratings.length;
+  } else {
+    return 0;
+  }
```

Comment:
If a function only has two return possibilities, there is no need to use the `else` block around the non-conditional return. Make it as a default return instead.
```
if (product) {
// statements
const total = ratings.reduce...
return total
}

return 0;
```